lib/Controller/BotController.php

```diff
@@ -107,16 +100,40 @@ public function sendMessage(string $token, string $message, string $referenceId
 					$botAttempt->getBotServer()->getSecret(),
 					$message
 				);
-				$bot = $botAttempt;
-				break;
+				return $botAttempt;
 			} catch (UnauthorizedException) {
 			}
 		}
 
-		if (!$bot instanceof Bot) {
-			$this->logger->debug('No valid Bot entry found');
-			$response = new DataResponse([], Http::STATUS_UNAUTHORIZED);
-			$response->throttle(['action' => 'bot']);
+		$this->logger->debug('No valid Bot entry found');
+		throw new \InvalidArgumentException('No valid Bot entry found', Http::STATUS_UNAUTHORIZED);
+	}
+
+	/**
+	 * Sends a new chat message to the given room.
+	 *
+	 * The author and timestamp are automatically set to the current user/guest
+	 * and time.
+	 *
+	 * @param string $token conversation token
+	 * @param string $message the message to send
+	 * @param string $referenceId for the message to be able to later identify it again
+	 * @param int $replyTo Parent id which this message is a reply to
+	 * @param bool $silent If sent silent the chat message will not create any notifications
+	 * @return DataResponse the status code is "201 Created" if successful, and
+	 *         "404 Not found" if the room or session for a guest user was not
+	 *         found".
+	 */
+	#[BruteForceProtection(action: 'bot')]
+	#[PublicPage]
+	public function sendMessage(string $token, string $message, string $referenceId = '', int $replyTo = 0, bool $silent = false): DataResponse {
+		try {
+			$bot = $this->getBotFromHeaders($token, $message);
+		} catch (\InvalidArgumentException $e) {
+			$response = new DataResponse([], $e->getCode());
+			if ($e->getCode() === Http::STATUS_UNAUTHORIZED) {
+				$response->throttle(['action' => 'bot']);
+			}
 			return $response;
```

Comment:
I am looking for a way to avoid repetition here and keep the same logic in one place. Maybe a wrapper like this:
```php
private function handleInvalidArgumentException(\InvalidArgumentException $e) {
	$response = new DataResponse([], $e->getCode());
	if ($e->getCode() === Http::STATUS_UNAUTHORIZED) {
		$response->throttle(['action' => 'bot']);
	}
	return $response;
}
```

Block from l. 130 to l. 138 is used several times. I wonder we could extract it in a method, but then we will have something like: 
```php
$bot = $this->loadBot($token, $message);
if ($bot instanceof DataResponse) {
	return $bot;
}
```
And I don't like it ^^

Reply:
yeah, I don't like mixed return types, that's why I went for the throwing instead.